### PR TITLE
Implementation factory support in ServiceCollection

### PR DESCRIPTION
**Pull request by jazzman55**
## Description

- Added _ImplementationFactory_ property to _ServiceDescriptor_ with delegate type: _ImplementationFactoryDelegate_.
- Added new constructor in _ServiceDescriptor_ with support for the new property.
- Added new extension methods for _IServiceCollection_ (_AddTransient_, _AddSingleton_, _AddScoped_) with factory method support.
- Added support for factory method in _ServiceProviderEngine_.
- Added new unit tests in _DependencyInjectionTests_.

## Motivation and Context

- Resolves nanoFramework/Home#1417

## How Has This Been Tested?

- unit tests

## Types of changes

- [ ] Improvement (non-breaking change that improves a feature, code or algorithm)
- [ ] Bug fix (non-breaking change which fixes an issue with code or algorithm)
- [x] New feature (non-breaking change which adds functionality to code)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Config and build (change in the configuration and build system, has no impact on code or features)
- [ ] Dependencies (update dependencies and changes associated, has no impact on code or features)
- [ ] Unit Tests (add new Unit Test(s) or improved existing one(s), has no impact on code or features)
- [ ] Documentation (changes or updates in the documentation, has no impact on code or features)

## Checklist:

- [x] My code follows the code style of this project (only if there are changes in source code).
- [ ] My changes require an update to the documentation (there are changes that require the docs website to be updated).
- [ ] I have updated the documentation accordingly (the changes require an update on the docs in this repo).
- [x] I have read the [CONTRIBUTING](https://github.com/nanoframework/.github/blob/main/CONTRIBUTING.md) document.
- [x] I have tested everything locally and all new and existing tests passed (only if there are changes in source code).
- [x] I have added new tests to cover my changes.

### CODE REVIEW

File: nanoFramework.DependencyInjection/Microsoft/Extensions/DependencyInjection/ServiceCollectionServiceExtensions.cs [hunk 1]
```diff
@@ -80,6 +80,30 @@ public static IServiceCollection AddSingleton(this IServiceCollection services,
...
+        /// with the specified <paramref name="factory"/> to the
+        /// specified <see cref="IServiceCollection"/>.
+        /// </summary>
+        /// <param name="services">The <see cref="IServiceCollection"/> to add the service to.</param>
+        /// <param name="serviceType">The type of the service to register.</param>
+        /// <param name="factory">A factory used for creating service instances.</param>
+        /// <returns>A reference to this instance after the operation has completed.</returns>
+        /// <seealso cref="ServiceLifetime.Singleton"/>
+        /// <exception cref="ArgumentNullException"><paramref name="services"/> can't be <see langword="null"/>.</exception>
+        public static IServiceCollection AddSingleton(this IServiceCollection services, Type serviceType, ImplementationFactoryDelegate factory)
```
Comment (Ellerbach) [inline nanoFramework.DependencyInjection/Microsoft/Extensions/DependencyInjection/ServiceCollectionServiceExtensions.cs:94]: Stupid question: is this method aligned with the .NET one?

File: nanoFramework.DependencyInjection/Microsoft/Extensions/DependencyInjection/ServiceCollectionServiceExtensions.cs [hunk 2]
```diff
@@ -123,6 +147,30 @@ public static IServiceCollection AddTransient(this IServiceCollection services,
...
+        /// with the specified <paramref name="factory"/> to the
+        /// specified <see cref="IServiceCollection"/>.
+        /// </summary>
+        /// <param name="services">The <see cref="IServiceCollection"/> to add the service to.</param>
+        /// <param name="serviceType">The type of the service to register.</param>
+        /// <param name="factory">A factory used for creating service instances.</param>
+        /// <returns>A reference to this instance after the operation has completed.</returns>
+        /// <seealso cref="ServiceLifetime.Transient"/>
+        /// <exception cref="ArgumentNullException"><paramref name="services"/> can't be <see langword="null"/>.</exception>
+        public static IServiceCollection AddTransient(this IServiceCollection services, Type serviceType, ImplementationFactoryDelegate factory)
```
Comment (Ellerbach) [inline nanoFramework.DependencyInjection/Microsoft/Extensions/DependencyInjection/ServiceCollectionServiceExtensions.cs:161]: And this one as well?

File: nanoFramework.DependencyInjection/Microsoft/Extensions/DependencyInjection/ServiceCollectionServiceExtensions.cs (hunk 1 above)
Comment (jazzman55) [reply nanoFramework.DependencyInjection/Microsoft/Extensions/DependencyInjection/ServiceCollectionServiceExtensions.cs]: Good point! I updated methods & comments.

**Review by Ellerbach — APPROVED** [review]
With better alignment, this looks all good :-)

**Comment by jazzman55** [comment]
@bytewizer , any thoughts?

**Comment by torbacz** [comment]
@Ellerbach @bytewizer 
It's waiting for 2 weeks, can you speed up with merge?

**Comment by Ellerbach** [comment]
> It's waiting for 2 weeks, can you speed up with merge?

We can but still have an issue with nuget. So if I merge now, it won't be available. So I4ll wait for the nuget issue to be resolved and then I'll merge.

**Pull request merged**